[RFC] #1999 - Add Targets.vim tutorial

nwneisen opened the pull request:
Add a tutorial with description, useful keys, and interactive examples
to teach users about Target.vim.

nwneisen commented:
Here is my work for issue #1999.  I ran into some trouble when I squashed my commits but I believe everything is straightened out now.  

I'm really happy with how this turned out.  I think the only changes I might look into more are adding a horizontal divider between the note keys and pausing after the user enters a correct position.

Looking forward to your feedback.

bryphe commented on browser/src/Services/Learning/Tutorial/Tutorials/TargetsVimPluginTutorial.tsx:
```diff
@@ -0,0 +1,142 @@
+                new Stages.SetBufferStage([Line5a, Line5b, Line5c]),
+            ),
+            new Stages.SetCursorPositionStage(0, 0),
+            new Stages.MoveToGoalStage("Use 'dina' to delete the next programming argument", 1, 31),
```
I noticed after the delete it moved quickly to the next stage (so I wasn't quite able to see the impact of my change) - is this what you referred to about adding a pause? I think that might be helpful, just to give the user some feedback about what the delete operation did.

bryphe commented on browser/src/Services/Learning/Tutorial/Tutorials/TargetsVimPluginTutorial.tsx:
```diff
@@ -0,0 +1,142 @@
+            new Stages.WaitForModeStage("Ready to type.  Go back to 'normal' mode", "normal"),
+            new Stages.SetCursorPositionStage(2, 7),
+            new Stages.MoveToGoalStage("Use 'cAn\"' to select around the double quotes", 2, 26),
+            new Stages.WaitForModeStage("Ready to type.  Go back to 'normal' mode", "normal"),
```
I realized at first when I got to this step - I wasn't sure if I was  supposed to do anything - I think the 'Ready to type.' was a bit ambigious (for me). We could consider making it a bit more explicit, like:
- "Press ESC to go back to 'normal' mode."

Or perhaps just:
- "Go back to 'normal' mode"


Or have them actually enter some text:
- "Type 'foo'."
- " Press ESC to go back to normal mode."

bryphe commented:
Hi @nwneisen , thank you for submitting the PR! Just played through the tutorial, it's great 👍 I like that you attributed the author in the body of the tutorial!

Just left some minor feedback in the comments - let me know what you think.

I also like the shorter version of some the commands:
- `ci"`, `ci'`, `ci(` - I wonder if it is worth calling out those too? (I think technically `ci"` and `ci'` are not part of the plugin, though - so perhaps a different tutorial is the proper place for those).

Overall the change looks good though, really appreciate it @nwneisen !

bryphe commented:
BTW, I noticed I'm starting to use `cin` more now thanks to this tutorial! 👍

nwneisen commented:
@bryphe I addressed your comments from the review.  I also added a section to show the difference between `ci` and `cin`.  Let me know what you think!

bryphe commented on browser/src/Services/Learning/Tutorial/Tutorials/TargetsVimPluginTutorial.tsx:
```diff
@@ -42,25 +42,33 @@ export class TargetsVimPluginTutorial implements ITutorial {
+                new Stages.SetBufferStage([Line2a, Line2b, Line2c, Line2d]),
             ),
             new Stages.SetCursorPositionStage(0, 7),
             new Stages.MoveToGoalStage("Use 'cin(' to move inside the paranthesis", 0, 40),
```
Minor, but I just realized `paranthesis` was misspelled a couple of places - should be `parenthesis`.

nwneisen replied:
Good catch.

bryphe commented on browser/src/Services/Learning/Tutorial/Tutorials/TargetsVimPluginTutorial.tsx:
```diff
@@ -42,25 +42,33 @@ export class TargetsVimPluginTutorial implements ITutorial {
-            new Stages.WaitForModeStage("Ready to type.  Go back to 'normal' mode", "normal"),
+            new Stages.WaitForModeStage("Press ESC to go back to normal mode", "normal"),
+            new Stages.SetCursorPositionStage(3, 34),
+            new Stages.MoveToGoalStage("Use 'ca(' to select inside the current paranthesis", 3, 32),
```
Is this one supposed to be `ci(` instead of `ca(`?

nwneisen replied:
I had the user continue to use ca( so that there wasn't a pair of parenthesis afterwards.  This will cause the next ci( to select the same pair of parenthesis.  I could change it and add a step to have the user move the cursor to the right and really drive the point of it selecting the current pair.

bryphe commented on browser/src/Services/Learning/Tutorial/Tutorials/TargetsVimPluginTutorial.tsx:
```diff
@@ -42,25 +42,33 @@ export class TargetsVimPluginTutorial implements ITutorial {
+            new Stages.SetCursorPositionStage(3, 34),
+            new Stages.MoveToGoalStage("Use 'ca(' to select inside the current paranthesis", 3, 32),
+            new Stages.WaitForModeStage("Press ESC to go back to normal mode", "normal"),
+            new Stages.MoveToGoalStage("Use 'ca(' to select the next paranthesis", 3, 40),
```
I realized for this line, and line 68, the word 'select' is a bit confusing (it's overloaded in Vim, since there is a `select mode`). Perhaps we could replace 'select' with 'change'?

nwneisen replied:
Good idea.  I will update anywhere using 'ci(' to 'change' and anywhere using 'di(' to delete.  I think associating the words for each shortcut will help to remember them.

bryphe commented:
Thanks a lot for making the changes, @nwneisen ! Overall looks good to me. Just had a couple other pieces of minor feedback - let me know what you think.

nwneisen commented:
Here are the vocabulary adjustments.  On top of doing 'c' for 'change' and 'd' for 'delete', I also looked at the second character (iaIA) and tried to match the vocabulary used for these in the Targets.vim repo as well.  

I didn't make any change to go from 'ca(' to 'ci('.  I feel like entering two 'ca(' in a row shows the user that its behavior can change without any other input.
